compiler/noirc_evaluator/src/ssa/ir/dfg.rs

```diff
@@ -497,9 +497,22 @@ impl DataFlowGraph {
         }
     }
 
-    /// True if the given ValueId refers to a constant value
+    /// True if the given ValueId refers to a (recursively) constant value
     pub(crate) fn is_constant(&self, argument: ValueId) -> bool {
-        !matches!(&self[self.resolve(argument)], Value::Instruction { .. } | Value::Param { .. })
+        match &self[self.resolve(argument)] {
+            Value::Instruction { .. } | Value::Param { .. } => false,
+            Value::Array { array, .. } => array.iter().all(|element| self.is_constant(*element)),
+            _ => true,
+        }
+    }
+
+    /// True if `is_constant` and if a NumericConstant, then is non-zero
+    pub(crate) fn is_constant_and_truthy(&self, argument: ValueId) -> bool {
+        self.is_constant(argument)
+            && match &self[self.resolve(argument)] {
+                Value::NumericConstant { constant, .. } => !constant.is_zero(),
+                _ => true,
```

Comment:
If we have a `Value::Array` we don't care if the internal elements are constant and truthy? Only constant? Should we not fail in any case other than that a boolean? Based on the issue I thought static_assert was more trying to mock `assert` rather than `assert_constant`. Wouldn't we just want this actually:
```suggestion
        if let Some(constant) = self.get_numeric_constant(argument) {
             !constant.is_zero()
        } else {
             false
        }
```

Reply:
yes, was thinking about this last night. it was too broad so have added your change 👍

Reply:
Looks like I missed a brace in my suggestion. When you fix that up we can merge!

Reply:
The suggestion fails on `static_assert(true, "")`, so I've simplified it to 1) use `self.resolve` as before, and 2) only check for `true`, since all other cases should fail and the stdlib function only accepts `bool`.
```rust
match &self[self.resolve(argument)] {
    Value::NumericConstant { constant, .. } => !constant.is_zero(),
    _ => false,
}
```

Reply:
`get_numeric_constant` calls a match internally which uses `self.resolve` 🤔

Reply:
reverted this piece: I was wrong and the issue was with how it was being called when evaluating `static_assert`